rascaline-c-api/src/logging.rs

```diff
@@ -0,0 +1,105 @@
+pub const RASCAL_LOG_LEVEL_INFO: i32 = 3;
+
+/// The "debug" level designates lower priority information
+///
+/// By default, log messages at this level are disabled in release mode, and
+/// enabled in debug mode.
+pub const RASCAL_LOG_LEVEL_DEBUG: i32 = 4;
+
+/// The "trace" level designates very low priority, often extremely verbose,
+/// information.
+///
+/// By default, rascaline disable this level, you can enable it by editing the
+/// code.
+pub const RASCAL_LOG_LEVEL_TRACE: i32 = 5;
+
+/// Callback function type for rascaline logging system. Such functions are
+/// called when a log event is emitted in the code.
+///
+/// The first argument is the log level, one of `RASCAL_LOG_LEVEL_ERROR`,
+/// `RASCAL_LOG_LEVEL_WARN` `RASCAL_LOG_LEVEL_INFO`, `RASCAL_LOG_LEVEL_DEBUG`,
+/// or `RASCAL_LOG_LEVEL_TRACE`. The second argument is a NULL-terminated string
+/// containing the message associated with the log event.
+#[allow(non_camel_case_types)]
+pub type rascal_logging_callback_t = Option<unsafe extern fn(level: i32, message: *const std::os::raw::c_char)>;
+
+// Mutex cannot use rust static
+// see https://stackoverflow.com/a/27826181
+lazy_static! {
+    static ref GLOBAL_CALLBACK: Mutex<rascal_logging_callback_t> = Mutex::new(None);
+}
+
+/// Implementation of `log::Log` that forward all log messages to the global
+/// `rascal_logging_callback_t`.
+struct RascalLogger;
+
+/// Set the given ``callback`` function as the global logging callback. This
+/// function will be called on all log events. If a logging callback was already
+/// set, it is replaced by the new one.
+#[no_mangle]
+pub unsafe extern fn rascal_set_logging_callback(callback: rascal_logging_callback_t) -> rascal_status_t {
+    catch_unwind(|| {
+        *GLOBAL_CALLBACK.lock().expect("mutex was poisoned") = callback;
+        // we allow multiple sets of logger, therefore the result will be ignored
+        let _ = log::set_boxed_logger(Box::new(RascalLogger));
+
+        if cfg!(debug_assertions) {
+            log::set_max_level(log::LevelFilter::Debug);
+        } else {
+            log::set_max_level(log::LevelFilter::Info);
+        }
+
+        Ok(())
+    })
+}
+
+
+impl log::Log for RascalLogger {
+    fn enabled(&self, _: &Metadata) -> bool {
+       return true;
+    }
+
+    fn log(&self, record: &Record) {
+        if self.enabled(record.metadata()) {
+            let message = format!("{} -- {}", record.target(), record.args());
+            let message_cstr = CString::new(message).unwrap();
+            unsafe {
+                match *(GLOBAL_CALLBACK.lock().expect("mutex was poisoned")) {
+                    Some(callback) => callback(record.level() as i32, message_cstr.as_ptr()),
+                    None => unreachable!("missing callback but RascalLogger is set as the global logger"),
+                }
+            }
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
```

Comment:
what does this do?

Reply:
Checking that the values exported above actually match the values from `log` (in case the values from log change at some point in the future, or we want to use a separate logging facade). Should I add a small comment explaining this?

Reply:
Oh right, I though you asked about the test module =). `use super::*;` is equivalent to python `import * from ..`. Like Python, using wildcard imports is usually discouraged, except in this exact situation: importing everything from the parent module to write corresponding tests. 

Without this, I could have written `use super::RASCAL_LOG_LEVEL_DEBUG` (similar to `from .. import RASCAL_LOG_LEVEL_DEBUG)`, or use the full access path from the crate root: `use crate::logging::RASCAL_LOG_LEVEL_DEBUG;` (similar to `from rascaline.logging import RASCAL_LOG_LEVEL_DEBUG`).